File: java/src/test/java/datastructures/LinkedListTest.java
package datastructures;

import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.*;

public class LinkedListTest {

    @Test
    public void shouldAddFrontOfList() {
        final var linkedList = new LinkedList<Integer>();
        linkedList.addFront(1);
        linkedList.addFront(2);
        linkedList.addFront(3);

        assertThat(linkedList.getFirst(), is(3));
        assertThat(linkedList.getLast(), is(1));
    }

    @Test
    public void shouldAddBackOfList() {
        final var linkedList = new LinkedList<Integer>();
        linkedList.addBack(1);

        assertThat(linkedList.getFirst(), is(1));
        assertThat(linkedList.getLast(), is(1));
    }

    @Test
    public void shouldGetSize() {
        final var linkedList = new LinkedList<String>();
        assertThat(linkedList.getSize(), is(0));
        assertThat(linkedList.count(), is(0));

        linkedList.addFront("add1");
        assertThat(linkedList.getSize(), is(1));
        assertThat(linkedList.count(), is(1));

        linkedList.addFront("add2");
        assertThat(linkedList.getSize(), is(2));
        assertThat(linkedList.count(), is(2));
    }

    @Test
    public void shouldClearTheList() {
        final var linkedList = new LinkedList<Integer>();
        linkedList.addFront(1);
        linkedList.addFront(2);
        linkedList.addFront(3);

        linkedList.clear();

        assertThat(linkedList.getSize(), is(0));
    }

    @Test
    public void shouldDeleteHeadFromListByValue() {
        final var linkedList = new LinkedList<Integer>();
        linkedList.addFront(1);
        linkedList.addFront(2);
        linkedList.addFront(3);

        linkedList.delete(3);

        assertThat(linkedList.getSize(), is(2));
        assertThat(linkedList.getFirst(), is(2));
        assertThat(linkedList.getLast(), is(1));
    }

    @Test
    public void shouldDeleteFromListByValue() {
        final var linkedList = new LinkedList<Integer>();
        linkedList.addFront(1);
        linkedList.addFront(2);
        linkedList.addFront(3);

        linkedList.delete(2);

        assertThat(linkedList.getSize(), is(2));
        assertThat(linkedList.getFirst(), is(3));
        assertThat(linkedList.getLast(), is(1));
    }


    @Test
    public void shouldDeleteTailFromListByValue() {
        final var linkedList = new LinkedList<Integer>();
        linkedList.addFront(1);
        linkedList.addFront(2);
        linkedList.addFront(3);

        linkedList.delete(1);

        assertThat(linkedList.getSize(), is(2));
        assertThat(linkedList.getFirst(), is(3));
        assertThat(linkedList.getLast(), is(2));
    }
}

File: kotlin/src/test/kotlin/datastructures/LinkedListTest.kt
package datastructures

import org.assertj.core.api.Assertions.assertThat
import org.junit.Ignore
import org.junit.Test

class LinkedListTest {

    @Test
    fun `Should Add To Front`() {
        val list = LinkedList<String>()

        assertThat(list.isEmpty()).isTrue()

        list.addFront("add1")
        list.addFront("add2")

        assertThat(list.size()).isEqualTo(2)
    }

    @Test
    fun `Should Get The First And Last Element`() {
        val list = LinkedList<Int>()
        list.addFront(1)
        list.addFront(2)

        assertThat(list.size()).isEqualTo(2)
        assertThat(list.getLast()).isEqualTo(1)
        assertThat(list.getFirst()).isEqualTo(2)
    }

    @Test
    fun `Should Add To Back`() {
        val list = LinkedList<String>()
        list.addFront("head")

        list.addBack("tail")

        assertThat(list.size()).isEqualTo(2)
        assertThat(list.getLast()).isEqualTo("tail")
        assertThat(list.getFirst()).isEqualTo("head")
    }

    @Test
    fun `Should Delete from List`() {
        val list = LinkedList<Any>()
        list.addFront("add1")
        list.addFront(3)
        list.addFront("add2")

        list.delete(3)

        assertThat(list.size()).isEqualTo(2)
        assertThat(list.getLast()).isEqualTo("add1")
        assertThat(list.getFirst()).isEqualTo("add2")

    }

    @Test
    fun `Should Delete HEAD from List with size 1`() {
        val list = LinkedList<Any>()
        list.addFront("head")

        list.delete("head")

        assertThat(list.size()).isEqualTo(0)
    }

    @Test
    fun `Should Delete HEAD from List`() {
        val list = LinkedList<String>()
        list.addFront("add1")
        list.addFront("add2")
        list.addFront("add3")

        list.delete("add3")

        assertThat(list.size()).isEqualTo(2)
        assertThat(list.getLast()).isEqualTo("add1")
        assertThat(list.getFirst()).isEqualTo("add2")
    }

    @Test
    fun `Should Delete TAIL from List`() {
        val list = LinkedList<String>()
        list.addFront("add1")
        list.addFront("add2")
        list.addFront("add3")

        list.delete("add1")

        assertThat(list.size()).isEqualTo(2)
        assertThat(list.getLast()).isEqualTo("add2")
        assertThat(list.getFirst()).isEqualTo("add3")
    }
}